webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala

```diff
@@ -218,33 +214,168 @@ class TechniqueWriter (
 
 trait AgentSpecificTechniqueWriter {
 
-  def writeAgentFile( technique : Technique, methods : Map[BundleName, GenericMethod] ) : IOResult[Option[String]]
+  def writeAgentFiles( technique : Technique, methods : Map[BundleName, GenericMethod] ) : IOResult[Seq[String]]
 
   def agentMetadata ( technique : Technique, methods : Map[BundleName, GenericMethod] ) : PureResult[NodeSeq]
 }
 
-class ClassicTechniqueWriter extends AgentSpecificTechniqueWriter {
+class ClassicTechniqueWriter(basePath : String) extends AgentSpecificTechniqueWriter {
+
+  // We need to add a reporting bundle for this method to generate a na report for any method with a condition != any/cfengine (which ~= true
+  def methodNeedReporting(method : MethodCall) =  method.condition != "any" && method.condition != "cfengine-community"
+  def needReportingBundle(technique : Technique) = technique.methodCalls.exists(methodNeedReporting)
+
+  def canonifyCondition(methodCall: MethodCall) = {
+    methodCall.condition.replaceAll("""(\$\{[^\}]*})""","""",canonify("$1"),"""")
+  }
+
+  // regex to match double quote characters not preceded by a backslash, and backslash not preceded by backslash or not followed by a backslash or a quote (simple or double)
+  def escapeCFEngineString(value : String ) = value.replaceAll("""(?<!\\)"""", """\\"""").replaceAll("""(?<!\\)\\(?!(\\|"|'))""","""\\\\""")
```

Comment:
This seems very brittle and a good candidate for a parser. We should create an issue for that.